### PR TITLE
Merge my wire addons into wire

### PR DESCRIPTION
**Pull request by stepa2**
This PR includes two my addons:

- [Improved RT Cameras](https://github.com/stepa2/improved-rt-cameras) [(Workshop)](https://steamcommunity.com/sharedfiles/filedetails/?id=2566118330)
Adds working wire-compatible render target cameras
- ~~[Expression 2 Entity Manipulation](https://github.com/stepa2/e2_entitymanipulation) [(Workshop)](https://steamcommunity.com/sharedfiles/filedetails/?id=2510195508)
Allows spawning engine entities, changing keyvalues and datadesc, firing inputs and hooking outputs. **Highly unsafe**, but powerful.~~

After (if) this PR will be merged, I plan to archive original repositories and remove workshop pages.

### CODE REVIEW

File: lua/entities/gmod_wire_expression2/core/custom/entmanipulation.lua [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+	
+	if !(ply:IsAdmin() or NoRestriction:GetBool() or isOwner(self, ent) or ent:MapCreationID() ~= -1) then 
+		ply:ChatPrint( "Entity:entFire() or Entity:entKVSet() error: You are only allowed to manipulate your own or map-created entities." ) 
+		return false
+	end
+	
+	return true
+end
+
+local E2_TABLE_DEFAULT = {n={},ntypes={},s={},stypes={},size=0}
```
Comment (Vurv78) [inline lua/entities/gmod_wire_expression2/core/custom/entmanipulation.lua:25]: You should be using ``E2Lib.newE2Table``, not this and constantly cloning

**Review by Vurv78 — CHANGES_REQUESTED** [review]
Small reviews from just glancing at this.
I think adding ``e:fire`` and ``e:setkeyvalues`` in any form, even with a whitelist, is an awful idea. You even acknowledge how badly this could be abused.  A lot of servers just randomly turn on every E2 core in wiremod as well.

Maybe this could make it into wire extras since it's more lenient, but right now I don't support this (dunno about the RT cameras)

File: lua/entities/gmod_wire_expression2/core/custom/entmanipulation.lua [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+
+	name = "e2_name_"..this:GetClass().."_"..tostring(this:GetCreationID())
+	this:SetName(name)
+
+	return name
+end
+
+__e2setcost(2)
+
+e2function table entity:entKVsGet()
```
[the remark this answers is not in the document]
Comment (stepa2) [reply lua/entities/gmod_wire_expression2/core/custom/entmanipulation.lua:173]: It I checked entity keyvalues, inputs, input values and datadesc values for validness, I could create three safety profiles:
1. Fully safe: players only can spawn whitelisted entities, fire whitelisted inputs etc.
2. Safe for non-admins: usual players can only use whitelisted functionality, admins can do everything
3. Unsafe mode: everyone can do everything

**Comment by stepa2** [comment]
It I checked entity keyvalues, inputs, input values and datadesc values for validness, I could create three safety profiles:

    Fully safe: players only can spawn whitelisted entities, fire whitelisted inputs etc.
    Safe for non-admins: usual players can only use whitelisted functionality, admins can do everything
    Unsafe mode: everyone can do everything

**Comment by stepa2** [comment]
I will not include EntManipulation (maybe later...)

**Comment by Divran** [comment]
Yeah the e2 entity extension could probably never be added. To wire extras, and disabled by default, maybe, but not in main wiremod.

Also I would recommend merging the two tools to spawn an RT camera and screen, into one tool where left click does one thing and right click the other.

**Comment by stepa2** [comment]
In future I would like to add support to rendering RT camera contents to EGP, and RT camera will not always be used with RT monitor. So I do not want to merge tools and treat RT camera and RT monitor as a complementary pair.

**Comment by stepa2** [comment]
Code is ready-to-merge

**Comment by Divran** [comment]
> In future I would like to add support to rendering RT camera contents to EGP, and RT camera will not always be used with RT monitor. So I do not want to merge tools and treat RT camera and RT monitor as a complementary pair.

Eh, I disagree that merging the tools implies you can't also use them for other things

**Comment by stepa2** [comment]
If I will merge tools, settings will take too many space.

изображение
изображение

**Comment by thegrb93** [comment]
Code looks fine to me

**Comment by Divran** [comment]
There's a function in the code called `SetMaterialParams` which doesn't seem to be defined anywhere. It's also not documented on the gmod wiki. What does this function do?

**Comment by RadioactvePixels** [comment]
RT Cameras cause a massive FPS drop... 2 FPS with 3 cameras/screens.

**Pull request merged**